include/swift/Localization/LocalizationFormat.h

```diff
@@ -173,16 +180,33 @@ class LocalizationProducer {
   virtual llvm::StringRef getMessageOr(swift::DiagID id,
                                        llvm::StringRef defaultMessage);
 
+  /// \returns a `SerializedLocalizationProducer` pointer if the serialized
+  /// diagnostics file available, otherwise returns a `YAMLLocalizationProducer`
+  /// if the `YAML` file is available. If both files aren't available returns a
+  /// `nullptr`.
+  static std::unique_ptr<LocalizationProducer>
+  producerFor(llvm::StringRef locale, llvm::StringRef path,
+              bool printDiagnosticNames);
+
   virtual ~LocalizationProducer() {}
 
 protected:
+  LocalizationProducerState getState() const;
+
+  /// Used to lazily initialize `LocalizationProducer`s.
+  /// \returns true if the producer is successfully initialized, false
+  /// otherwise.
+  virtual bool initializeImpl() = 0;
```

Comment:
@xedin I'm just not sure when `YAMLLocalizationProducer::initializeImpl` or `SerializedLocalizationProducer::initializeImpl` will ever fail and return false. I'm returning true in both implementations for now.

Reply:
I mean, the current state (before this PR) is that in both Serialized & YAML LocalizationProducers' constructors doesn't fail. Or at least we don't handle failures in the constructors.

Reply:
It might not matter now but APIs change so it's prudent to account for possibility of a failure.

Reply:
Makes sense, thank you!